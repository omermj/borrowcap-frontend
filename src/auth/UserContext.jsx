import React from "react";

/** Context to provide currentUser thoughout the app */

const UserContext = React.createContext();

export default UserContext;
