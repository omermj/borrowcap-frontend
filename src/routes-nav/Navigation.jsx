import { Link, NavLink } from "react-router-dom";
import { Container, Nav, Navbar, NavDropdown } from "react-bootstrap";
import UserContext from "../auth/UserContext";
import { useContext } from "react";

const Navigation = ({ logout }) => {
  const { currentUser } = useContext(UserContext);

  const BorrowerNav = () => {
    return (
      <Nav className="ms-auto">
        <Nav.Link href="/borrower/apply">Apply for Loan</Nav.Link>
        <NavDropdown title="Applications" id="nav-dropdown">
          <NavDropdown.Item href="/borrower/activerequests">
            Requests
          </NavDropdown.Item>
          <NavDropdown.Item href="/borrower/approvedloans">
            Approved
          </NavDropdown.Item>
        </NavDropdown>
        <Nav.Link href="/borrower/fundedloans">Loans</Nav.Link>
        <Nav.Link href="/borrower/wallet">Wallet</Nav.Link>
        <Nav.Link href="/borrower/profile">Profile</Nav.Link>
        <Nav.Link onClick={logout}>Logout {currentUser.username}</Nav.Link>
      </Nav>
    );
  };

  const InvestorNav = () => {
    return (
      <Nav className="ms-auto">
        <Nav.Link href="/investor/availableinvestments">Invest</Nav.Link>
        <Nav.Link href="/investor/pledges">Pledges</Nav.Link>
        <Nav.Link href="/investor/activeinvestments">Investments</Nav.Link>
        <Nav.Link href="/investor/wallet">Wallet</Nav.Link>
        <Nav.Link href="/investor/profile">Profile</Nav.Link>
        <Nav.Link onClick={logout}>Logout {currentUser.username}</Nav.Link>
      </Nav>
    );
  };

  const UnderwriterNav = () => {
    return (
      <Nav className="ms-auto">
        <Nav.Link href="/underwriter/requests">Requests</Nav.Link>
        <Nav.Link href="/underwriter/approved">Approved</Nav.Link>
        <Nav.Link href="/" onClick={async () => await logout()}>
          Logout {currentUser.username}
        </Nav.Link>
      </Nav>
    );
  };

  const LoggedOutNav = () => {
    return (
      <Nav className="ms-auto">
        <Nav.Link href="/signup">Signup</Nav.Link>
        <Nav.Link href="/login">Login</Nav.Link>
      </Nav>
    );
  };

  let navToDisplay = null;
  if (currentUser) {
    if (currentUser.roles.includes("admin")) navToDisplay = UnderwriterNav;
    else if (currentUser.roles.includes("borrower")) navToDisplay = BorrowerNav;
    else if (currentUser.roles.includes("investor")) navToDisplay = InvestorNav;
  } else navToDisplay = LoggedOutNav;

  return (
    <Navbar bg="dark" data-bs-theme="dark" className="mb-4">
      <Container>
        <Navbar.Brand href="/">BorrowCap</Navbar.Brand>
        <Navbar.Toggle aria-controls="basic-navbar-nav" />
        <Navbar.Collapse id="basic-navbar-nav">
          {navToDisplay()}
        </Navbar.Collapse>
      </Container>
    </Navbar>
  );
};

export default Navigation;
