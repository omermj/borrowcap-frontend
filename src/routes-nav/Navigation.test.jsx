import { render } from "@testing-library/react";
import Navigation from "./Navigation";
import { BrowserRouter } from "react-router-dom";
import UserContext from "../auth/UserContext";

// smoke test
it("renders without crashing", () => {
  render(
    <BrowserRouter>
      <UserContext.Provider value={{ currentUser: { username: "test" } }}>
        <Navigation />
      </UserContext.Provider>
    </BrowserRouter>
  );
});

// snapshot test
it("matches snapshot", () => {
  const { asFragment } = render(
    <BrowserRouter>
      <UserContext.Provider value={{ currentUser: { username: "test" } }}>
        <Navigation />
      </UserContext.Provider>
    </BrowserRouter>
  );
  expect(asFragment()).toMatchSnapshot();
});
