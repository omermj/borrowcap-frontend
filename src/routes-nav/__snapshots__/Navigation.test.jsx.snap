// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <nav
    class="navbar navbar-expand navbar-light bg-dark"
    data-bs-theme="dark"
  >
    <div
      class="container-fluid"
    >
      <a
        class="navbar-brand"
        href="/"
      >
        BorrowCap
      </a>
      <button
        aria-controls="basic-navbar-nav"
        aria-label="Toggle navigation"
        class="navbar-toggler collapsed"
        type="button"
      >
        <span
          class="navbar-toggler-icon"
        />
      </button>
      <div
        class="navbar-collapse collapse"
        id="basic-navbar-nav"
      >
        <div
          class="ms-auto  navbar-nav"
        >
          <a
            class="nav-link"
            href="/user/profile"
          >
             
            <span>
              Profile
            </span>
             
          </a>
          <div
            class="nav-item dropdown"
          >
            <a
              aria-expanded="false"
              class="dropdown-toggle nav-link"
              href="#"
              id="nav-dropdown"
              role="button"
              tabindex="0"
            >
              <span>
                 
                <i
                  class="fs-6 bi-person"
                />
              </span>
            </a>
          </div>
        </div>
      </div>
    </div>
  </nav>
</DocumentFragment>
`;
